src/main/scala/mesosphere/marathon/state/MarathonStore.scala

```diff
@@ -1,139 +1,75 @@
 package mesosphere.marathon.state
 
-import com.codahale.metrics.MetricRegistry.name
-import com.codahale.metrics.{ Histogram, MetricRegistry }
-import com.google.protobuf.InvalidProtocolBufferException
-import mesosphere.marathon.{ MarathonConf, StorageException }
-import mesosphere.util.{ BackToTheFuture, LockManager, ThreadPoolContext }
-import org.apache.mesos.state.State
+import com.codahale.metrics.{ MetricRegistry, Histogram }
+import com.codahale.metrics.MetricRegistry._
+import mesosphere.util.{ LockManager, ThreadPoolContext }
+import mesosphere.util.state.PersistentStore
 import org.slf4j.LoggerFactory
 
-import scala.collection.JavaConverters._
 import scala.concurrent.Future
-import scala.concurrent.duration._
-import scala.util.control.NonFatal
+import scala.reflect.ClassTag
 
 class MarathonStore[S <: MarathonState[_, S]](
-  conf: MarathonConf,
-  state: State,
-  registry: MetricRegistry,
-  newState: () => S,
-  prefix: String = "app:")(
-    implicit val timeout: BackToTheFuture.Timeout = BackToTheFuture.Timeout(Duration(conf.marathonStoreTimeout(),
-      MILLISECONDS)))
-    extends PersistenceStore[S] {
+    val store: PersistentStore,
+    registry: MetricRegistry,
+    newState: () => S,
+    prefix: String = "app:")(implicit ct: ClassTag[S]) extends EntityStore[S] {
 
-  import BackToTheFuture.futureToFutureOption
   import ThreadPoolContext.context
-
   private[this] val log = LoggerFactory.getLogger(getClass)
-  private[this] lazy val locks = LockManager[String]()
-
-  def contentClassName: String = newState().getClass.getSimpleName
 
-  protected[this] val bytesRead: Histogram = registry.histogram(name(getClass, contentClassName, "read-data-size"))
-  protected[this] val bytesWritten: Histogram = registry.histogram(name(getClass, contentClassName, "write-data-size"))
+  private[this] lazy val locks = LockManager[String]()
+  private[this] lazy val bytesRead: Histogram =
+    registry.histogram(name(getClass, ct.getClass.getSimpleName, "read-data-size"))
+  private[this] lazy val bytesWritten: Histogram =
+    registry.histogram(name(getClass, ct.getClass.getSimpleName, "write-data-size"))
 
   def fetch(key: String): Future[Option[S]] = {
-    state.fetch(prefix + key) map {
-      case Some(variable) =>
-        bytesRead.update(variable.value.length)
-        try {
-          Some(stateFromBytes(variable.value))
-        }
-        catch {
-          case e: InvalidProtocolBufferException =>
-            if (variable.value.nonEmpty) {
-              log.error(s"Failed to read $key, could not deserialize data (${variable.value.length} bytes).", e)
-            }
-            None
-        }
-      case None =>
-        throw new StorageException(s"Failed to read $key, does not exist, should have been created automatically.")
+    store.load(prefix + key).map {
+      _.map{ entity =>
+        bytesRead.update(entity.bytes.length)
+        stateFromBytes(entity.bytes)
+      }
```

Comment:
A recover that adds context and rethrows an exception would be nice.

For example, if serialization fails, it would be very nice to know for what kind of entity and for which ID.